File: bemkl_supervised_classification_variational_test.m
function prediction = bemkl_supervised_classification_variational_test(Km, state)
    N = size(Km, 2);
    P = size(Km, 3);

    prediction.G.mu = zeros(P, N);
    prediction.G.sigma = zeros(P, N);
    for m = 1:P
        prediction.G.mu(m, :) = state.a.mu' * Km(:, :, m);
        prediction.G.sigma(m, :) = state.parameters.sigma_g^2 + diag(Km(:, :, m)' * state.a.sigma * Km(:, :, m));
    end

    prediction.f.mu = [ones(1, N); prediction.G.mu]' * state.be.mu;
    prediction.f.sigma = 1 + diag([ones(1, N); prediction.G.mu]' * state.be.sigma * [ones(1, N); prediction.G.mu]);

    pos = 1 - normcdf((+state.parameters.margin - prediction.f.mu) ./ prediction.f.sigma);
    neg = normcdf((-state.parameters.margin - prediction.f.mu) ./ prediction.f.sigma);
    prediction.p = pos ./ (pos + neg);
end

File: bemkl_supervised_classification_variational_test.R
bemkl_supervised_classification_variational_test <- function(Km, state) {
  N <- dim(Km)[2]
  P <- dim(Km)[3]

  G <- list(mu = matrix(0, P, N), sigma = matrix(0, P, N))
  for (m in 1:P) {
    G$mu[m,] <- crossprod(state$a$mu, Km[,,m])
    G$sigma[m,] <- state$parameters$sigma_g^2 + diag(crossprod(Km[,,m], state$a$sigma) %*% Km[,,m])
  }
  
  f <- list(mu = matrix(0, N, 1), sigma = matrix(0, N, 1))
  f$mu <- crossprod(rbind(matrix(1, 1, N), G$mu), state$be$mu)
  f$sigma <- 1 + diag(crossprod(rbind(matrix(1, 1, N), G$mu), state$be$sigma) %*% rbind(matrix(1, 1, N), G$mu))

  pos <- 1 - pnorm((+state$parameters$margin - f$mu) / f$sigma)
  neg <- pnorm((-state$parameters$margin - f$mu) / f$sigma)
  p <- pos / (pos + neg)

  prediction <- list(G = G, f = f, p = p)
}
